### DISCUSSION TITLE
프레임워크 / 라이브러리 결정 방식 제안

### DISCUSSION THREAD

**Discussion by chadseok** [wanted-pre-onboarding-side-project/habit-tracker#29] [Ideas]
![framework/library-decision](https://user-images.githubusercontent.com/83488329/230767200-adbf5a22-28cd-4a98-9db1-e654c46f88a0.png)

### 저는 다음과 같이 4분면을 통해서 프레임워크 / 라이브러리를 결정하는 방식을 제안합니다. 

이 방식에서 고려할 사항은 2가지입니다
1. 프레임워크/라이브러리가 이번 프로젝트에서 얼마나 개발 효율성을 높여줄 것인가?
2. 프레임워크/라이브러리가 이번 프로젝트의 코드 로직에서 얼마나 관여할 것 인가?

프로젝트에서 프레임워크/라이브러리를 사용하면 개발자는 정말 필요한 로직에 집중할 수 있기 때문에 이 프레임워크/라이브러리가 없으면 정말 개발하기 힘들고 없으면 하기 힘들다하는 프레임워크/라이브러리는 설치하면 좋을 것 같습니다

그리고 프레임워크/라이브러리는 개발 중간에 바뀔 수도 있고 추가될 수 도 있습니다. 이 때 사용할 프레임워크/라이브러리가 프로젝트 코드 로직에 얼마나 관여하는지를 체크해야합니다. 관여를 많이 할 수록  프레임워크/라이브러리에 문제가 생기면 에러 핸들링이 힘들어집니다. 바꿀 코드도 많아집니다. 그래서 코드 로직에 덜 관여하는 프레임워크/라이브러리는 설치해도 좋을 것 같습니다

---

### 각 4분면의 색깔에 대해서 설명입니다

제가 말씀드린 기준은 개발효율성을 높여주고 코드로직에 덜 관여할수록 설치해도 좋은 라이브러리/프레임워크입니다. 그래서 그것에 따라 사분면의 색깔을 정하였습니다

- 1번 사분면에 포함이 되면 설치를 해도 됩니다. (초록색으로 표시하였습니다)
- 2,3번 사분면에 포함이 된다면 팀과 토론 후 설치해야 합니다. (노란색으로 표시하였습니다)
- 4번 사분면에 해당이 된다면 설치를 하지 않는 것이 좋습니다. (빨간색으로 표시하였습니다)

---

### 두 고려사항에 대한 우선순위입니다

저희 프로젝트는 오랫동안 리팩토링하면서 유지해나갈 프로젝트가 아니라고 생각이듭니다. 그래서 단시간에 프로젝트를 진행하기 위해서는 **개발효율성** 측면이 **코드관여성**보다 더 우선된다고 생각합니다.

따라서 개발효율성이 엄청나지만 코드관여성도 높은 경우에는 저희 프로젝트에서는 충분히 설치해도 된다고 생각할 수 있습니다

**Comment by jiyeon2** [comment]
ㅇ창환님의 제안 저는 좋습니다~~
라이브러리/프레임워크를 도입할 지, 하지 않을지 여부를 결정할 때
창환님이 제시해주신 개발효율성, 코드관여성 두 가지 기준이 유용하다고 생각합니다

특정 기능을 위한 라이브러리 도입은 결정했지만 여러 후보군이 있을 수 있습니다
예를 들어, 기업과제 중 데이터 시각화 라이브러리 선정 시 저희 조는 recharts, react-chartjs-2 등을 고려했습니다

'프레임워크/라이브러리에 문제가 생기면 에러 핸들링이 힘들어진다'는 문제를 최소화 하기 위해 
라이브러리 선정 시 해당 라이브러리가 안정된 상태인지
즉, 버전이 높거나 최근까지 유지보수가 진행되었는지를 추가 기준으로 고려하면 좋겠습니다!

**Comment by greyHairChooseLife** [comment]
저도 좋습니다.

앞으로 개발 진행하면서 추가적인 라이브러리 도입 제안 시 근거로 삼도록 하겠습니다. 지연님이 추가 해 주신 기준도 포함해서요!
